Add cross-attention support to Attention layer

Jackryd opened the pull request:
Modified the Attention function to allow for different Q, K, and V matrices to enable cross-attention functionality. This maintains backward compatibility with the previous implementation since when the key is not specified, it defaults to the query (Q=K for self-attention) and the same holds for the value matrix.

murrellb commented:
Thanks! Some requests:

1. Can we call in input arguments something other than query, key, and value, because those don't generally refer to the inputs? Maybe `x_key`, `x_query` etc?
2. Is there a good motivation for having a separate `value` input? It's dimension will have to match the keys (if I recall the convention properly), but I've only seen cases where keys/values are derived, via different projections, from the same tensor.
3. Can we handle the cross vs self via dispatch instead of type unions and conditionals? Ie. define the main function as cross attention, and have a call that only passes in a single x just call the cross attention function, repeating the x?

Jackryd commented:
I made some changes now. I had no specific reason for a separate `value` input so I've removed it so that the function just uses the `x_key` instead. I then also changed the self-attention function to use dispatch.
